Location Reassignment Phase 1/3 and partial 2/3

mkangia opened the pull request:
Rework of https://github.com/dimagi/commcare-hq/pull/26865
First three commits picks up relevant code from that PR with minor changes and then build on it

[Spec Doc](https://docs.google.com/document/d/1h6kTr9M0FumswP3-gr_84OLeUHRmn5Y10dCpSAS3Y9g/edit#heading=h.oc6msb3kr2ii)

PRs merged into this one,
https://github.com/dimagi/commcare-hq/pull/26983
https://github.com/dimagi/commcare-hq/pull/27012
https://github.com/dimagi/commcare-hq/pull/27056

##### SUMMARY
This implements the first phase of development for the location reassignment feature.
The first phase is to add the ability for HQ to share a request template for a change request which can be edited to upload back to HQ to validate the transitions to be done.

##### ENVIRONMENTS
ICDS, India, Staging

##### FEATURE FLAG
`LOCATION_REASSIGNMENT`

##### PRODUCT DESCRIPTION
Adds a simple UI, preview [here](https://github.com/dimagi/commcare-hq/pull/26892) or up to date working example on [staging](https://staging.commcarehq.org/a/ccqa/location_reassignment/)

mkangia committed:
init location search select

mkangia committed:
isort

mkangia committed:
add UI

Extracted from https://github.com/dimagi/commcare-hq/pull/26865

mkangia commented on corehq/apps/locations/static/locations/js/locations.js:
```diff
@@ -1,12 +1,11 @@
     'hqwebapp/js/initial_page_data',
-    'locations/js/utils',
     'locations/js/location_tree',
+    'locations/js/search',
```
@orangejenny this is a bit different from how i did it originally with just renaming but now i reordered and removed it entirely.
Infact i think it does not need to included at all because location_tree already includes it.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/d9b9dd93c5edcd9585a602d0e7589041bddb7c6e

mkangia committed:
generate content for download

mkangia committed:
use Move for move operation

mkangia committed:
rename and use const

mkangia committed:
parse request

mkangia committed:
dump parsed request

mkangia committed:
remove unused import

mkangia committed:
remove redundant location type

mkangia committed:
add validations for workbook

mkangia committed:
add tests for parser

mkangia committed:
merge master

mkangia commented:
@orangejenny @esoergel this is ready for review now.

mkangia committed:
remove unused imports

mkangia committed:
validate descendants archived

mkangia committed:
add utils to deprecate locations

mkangia committed:
update parser to perform updates

mkangia committed:
add UI

mkangia committed:
fix argument

mkangia committed:
fix error message

orangejenny commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+        # setup details for the main location and all related locations
+        self._location_details_by_location_id = {}
+        for location in self._locations():
+            self._location_details_by_location_id[location.location_id] = {
```
This would be a good place to use a namedtuple.

orangejenny commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+        # memoized static content for each location
+        location_details = self._location_details_by_location_id[location_id]
+        location_parent_id = location_details['parent_location_id']
+        location_content = {
```
This would be another good namedtuple.

esoergel commented on custom/icds/location_reassignment/dumper.py:
```diff
@@ -0,0 +1,102 @@
+        export_raw(headers, rows, stream)
+        return stream
+
+    def _setup_site_codes(self, transitions):
```
I have the same sorts of comments about this class - instance variables are being defined inside methods and referenced elsewhere, so there's an unstated order in which things must be called.

mkangia replied:
there is only one public method which controls that order so that would keep that clear?

orangejenny commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
 class Download(object):
-    def __init__(self, domain, parent_location_id, leaf_location_type):
-        pass
+    def __init__(self, location_id):
```
Why was domain removed? Generally I think it's good not to assume ids are globally unique.

mkangia replied:
`location_id` are marked as unique is model/db, so should be safe here?

orangejenny replied:
They are right now, and I don't think they're likely to change, but I also wouldn't want to go through something like https://github.com/dimagi/commcare-hq/issues/25718 again.

mkangia replied:
That would be some deep architectural change, and i hope we would check all usages before making them non unique. Also there are other places that filter just with location_id.

esoergel replied:
I'd bank on ids being globally unique (since `site_code` is the "unique on the domain" thing).  Still, I do like having `domain` in there to make sure the user doesn't inadvertently (or maliciously) change something on another domain.  On ICDS in particular where there are test domains on the same environment, I could see someone getting mixed up and copying an ID from the production domain into an upload intended for the staging domain.

mkangia replied:
Only location load happening here is via location_id and then fetching all locations related to it.
The location is chosen via UI where someone would see locations for domain only.
Happy to add a check for domain as well.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/d6144a7041fe316a6ec0135eb1c11ac3e854f777

orangejenny commented on custom/icds/location_reassignment/dumper.py:
```diff
@@ -1,10 +1,102 @@
+                    [new_site_codes.extend(to_site_codes) for to_site_codes in list(details.values())]
+                else:
+                    new_site_codes.extend(list(details.keys()))
+        return new_site_codes
```
This seems like an easy place to raise an error if there are duplicates, just by comparing `len(new_site_codes)` with `len(set(new_site_codes))`

mkangia replied:
validations would be done via `Parser` and it checks for multiple operations to single site code, so it would get caught there itself.

orangejenny commented on custom/icds/templates/icds/partials/location_reassignment.html:
```diff
@@ -14,22 +14,4 @@
-    </p>
-    <div class="row">
-      <div class="col-sm-4">
-        <select class="form-control hqwebapp-select2" data-bind="value: location_type">
```
I think you can also remove `hqwebapp/js/widgets` from the js file.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/0a4c463c2b86491b71e932f471b9888b619830a5

orangejenny commented on custom/icds/location_reassignment/views.py:
```diff
@@ -63,8 +64,17 @@ def post(self, request, *args, **kwargs):
     def _workbook_is_valid(self, workbook):
-        # ToDo: Add necessary checks for workbook
-        return []
+        # ensure worksheets present and with titles as the location type codes
```
Should this also check that there are no missing tabs?

mkangia replied:
I had that earlier but i am currently keeping things flexible because i might be asked to not need all sheets and just relevant ones/ones that change.

orangejenny reviewed (commented):
Looks good, my comments are pretty minor.

orangejenny commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+            row = deepcopy(location_content)
+            row['username'] = username
+            row['new_username'] = ''
+            row['request'] = ''
```
Nevermind, you update this in a later commit.

orangejenny commented on custom/icds/static/icds/js/location_reassignment.js:
```diff
@@ -0,0 +1,37 @@
+            return self;
+        };
+
+        var $content = $("#hq-content");
```
This was me, it really is the smallest scope that contains both the download and upload sections of the page. Because this page uses HQ's standard bulk upload/download there's limited control.

orangejenny commented on custom/icds/static/icds/js/location_reassignment.js:
```diff
@@ -0,0 +1,37 @@
+
+        var $content = $("#hq-content");
+        $content.koApplyBindings(LocationReassignmentModel({
+            baseUrl: $content.find("#download_link").attr("href"),
```
This is also me, from https://github.com/dimagi/commcare-hq/pull/26892

There are a couple of other places that use similar code (locations import, app manager translations). It'd be good to rework the bulk uploader to be more friendly to javascript and other customization, but in the meantime I'd rather stick with the existing pattern, even though it's a janky pattern.

mkangia committed:
remove including js, included via location_tree

mkangia commented on custom/icds/location_reassignment/views.py:
```diff
@@ -0,0 +1,103 @@
+
+    def _generate_response(self, transitions):
+        response_file = Dumper().dump(transitions)
+        response_file.seek(0)
```
hmm, i think it's a good thing for any code receiving a file stream to seek if needed to be sure about reading file correctly. Though i am fine either way.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/e67455340132f3abdefc91b18e9657c451191a5a

mkangia commented on custom/icds/location_reassignment/views.py:
```diff
@@ -0,0 +1,103 @@
+        return errors
+
+    def _generate_response(self, transitions):
+        response_file = Dumper().dump(transitions)
```
the classes have that context @esoergel

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -0,0 +1,134 @@
+                'location_type': location.location_type.code,
+                'assigned_users': [],
+            }
+        return self._location_details_by_location_id
```
that return was an oversight and was not needed.

I find it nice when steps are laid down, those are actually two different steps.
```
def dump(self):
    self._init_location_details()
    self._populate_assigned_users()
```
or something like
```
def dump(self):
    self._fetch_locations_content():

def _fetch_locations_content(self):
    self._init_location_details()
    self._populate_assigned_users()
```
which is kind of the same.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/4228c775bcc25dedf8635045fd619d67e3085f64

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -0,0 +1,134 @@
+    def _get_users(self):
+        location_ids = list(self._location_details_by_location_id.keys())
+        # ToDo: confirm that its good to use base_username instead of username from ES, something about
+        #  them not being analysed
```
Thank you :)

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/add67810d23aede73356d5e78b899d024840929f

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -0,0 +1,134 @@
+    def _populate_assigned_users(self):
+        # allot assign user details to each location
+        for user_detail in self._get_users():
+            username, assigned_location_ids = user_detail
```
thats embarassing, thanks

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/26951/commits/6ab443cc51759b2812884500202237cc1dd349ef

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -0,0 +1,134 @@
+            row['username'] = username
+            row['new_username'] = ''
+            row[OPERATION_COLUMN] = ''
+            rows[location_type].append(row)
```
Ya, it would use whatever the current value of those variables are. Do you think its better if they are sent as arguments?

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -0,0 +1,134 @@
+
+    @memoized
+    def _location_content(self, location_id):
+        # memoized static content for each location
```
ya location_details is what we get from HQ and location_content is the content for that location row in the excel sheet which includes empty column names.
Just keeping those two distinct.

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+            self._location_details_by_location_id[location.location_id] = {
+                'name': location.name,
+                'site_code': location.site_code,
+                'lgd_code': location.metadata.get('lgd_code', ''),
```
its something that ICDS uses, its a metadata/custom location field

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+                if location_id in self._location_details_by_location_id:
+                    self._location_details_by_location_id[location_id]['assigned_users'].append(username)
+
+    def _get_users(self):
```
Eventually we just need to iterate over what ES gives, so making that modification adds an extra iteration which does not change anything.

orangejenny replied:
I think it's clearer to have a dict than tuples. It wouldn't make the later iteration any more complicated, would it?

mkangia replied:
Updated in combination with other feedback in https://github.com/dimagi/commcare-hq/pull/26951/commits/51f3732559a99a4191e2aa37dc34b61186043f44

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+        for user_detail in self._get_users():
+            username, assigned_location_ids = user_detail
+            if not isinstance(assigned_location_ids, list):
+                assigned_location_ids = [assigned_location_ids]
```
`get_users` just pulls the users from ES, why should this be moved there?

orangejenny replied:
Because it would make `get_users` return a more consistent value, where all `assigned_location_ids` are lists. As is, any other caller of `get_users` would need to do this same check.

mkangia replied:
Aha, i see.

https://github.com/dimagi/commcare-hq/pull/26951/commits/51f3732559a99a4191e2aa37dc34b61186043f44

mkangia commented on custom/icds/location_reassignment/download.py:
```diff
@@ -1,10 +1,129 @@
+        return wb
+
+    def _create_rows(self):
+        def append_row(username):
```
not passing username would mean something like

```
            if location_details['assigned_users']:
                for username in location_details['assigned_users']:
                    append_row()
            else:
                username = ''
                append_row()
```
so this just avoids that `username = ''`

orangejenny replied:
What about passing them both as parameters, instead of neither?

mkangia replied:
Ya, that was my question on another place where @esoergel suggested the same, if passing arguments instead of reusing variables from the context of enclosing method which makes me rethink the purpose of having a method defined within a method itself.

orangejenny replied:
I admit I don't like nested functions in python, probably because I'm more used to seeing them in javascript, where the scoping rules are slightly different. For this code, my only legitimate complaint is the inconsistency in the two params being different.

mkangia committed:
stickler

mkangia committed:
seek earlier

mkangia committed:
use lazy

mkangia committed:
remove redundant return

mkangia committed:
confirmed

mkangia committed:
better iteration

mkangia committed:
remove redundant js

mkangia committed:
inherit from correct baseview

mkangia commented on custom/icds/location_reassignment/views.py:
```diff
@@ -0,0 +1,105 @@
+
+@method_decorator([toggles.LOCATION_REASSIGNMENT.required_decorator()], name='dispatch')
+@method_decorator(require_can_edit_locations, name='dispatch')
+class LocationReassignmentView(BaseDomainView):
```
https://github.com/dimagi/commcare-hq/pull/26951/commits/cd866ea1da6a7f158f81a608c5cf92f271ef2ce9

mkangia commented on custom/icds/location_reassignment/views.py:
```diff
@@ -0,0 +1,105 @@
+class LocationReassignmentView(BaseDomainView):
+    section_name = ugettext_lazy("Locations")
+
+    page_title = _('Location Reassignment')
```
https://github.com/dimagi/commcare-hq/pull/26951/commits/f10d3f383128f3d0a2927e883e23a43e89488e36

mkangia commented:
@esoergel  thanks for feedback, responded and updated.

@orangejenny going to leave named tuples change for a follow up PR unless its a blocker, responded and updated for rest.

orangejenny reviewed (approved):
@mkangia I made responses to some of your responses, but nothing is blocking from me.

mkangia committed:
consolidate location ids when fetch user details itself

mkangia commented:
Thanks @orangejenny , can you mark the ones were you are happy with the response/update as resolved?

mkangia committed:
Merge branch 'mk/location-reassignment-1' into mk/location-reassignment-2

mkangia committed:
call process directly

orangejenny commented:
@mkangia Marked the clearly resolved conversations as resolved. The ones that are still open aren't blockers. Added one comment that might be a bug.

orangejenny committed:
Merge branch 'master' into mk/location-reassignment-1

orangejenny commented on corehq/toggles.py:
```diff
@@ -1828,10 +1828,19 @@ def _commtrackify(domain_name, toggle_is_enabled):
     help_link='https://confluence.dimagi.com/display/ccinternal/Two-Stage+Mobile+Worker+Account+Creation',
 )
 
+LOCATION_REASSIGNMENT = StaticToggle(
```
Not worth changing now, but it's helpful to put new toggles in the middle of the file, not the end, to reduce merge conflicts.

mkangia committed:
fix iteration

mkangia committed:
extract processor and make async

mkangia committed:
no need for cached property

mkangia committed:
fix tests for parser changes

mkangia committed:
add tests for processor

mkangia committed:
fix text

mkangia committed:
fix early return

mkangia committed:
fix return

mkangia committed:
tests for models, processor and utils

mkangia committed:
raise exception if missing locations

mkangia committed:
filter on domain as well

mkangia committed:
fix argument order

mkangia committed:
Merge pull request #26983 from dimagi/mk/location-reassignment-2

Location Reassignment Phase 2

mkangia committed:
make workbook write only, resolves empty sheet in beginning

mkangia committed:
better error management

mkangia committed:
update other model validations to use error attr

mkangia committed:
notify user

mkangia committed:
fix tests

mkangia committed:
pass user email

mkangia committed:
handle the order of locations correctly

mkangia committed:
Fix check for metadata

mkangia committed:
add comment

mkangia committed:
add a dropdown for operation by adding datavalidation

mkangia committed:
get cases owned by location

mkangia committed:
assign a new validation because cant reuse

mkangia committed:
update columns and extract to const

mkangia committed:
use predefined const instead

mkangia committed:
Merge pull request #27012 from dimagi/mk/location-reassignment-3

Location Reassignment Phase 1 & 2: better errors and notify result to user

mkangia committed:
fetch all old location ids at once for case count queries

mkangia committed:
add tests for dumper

mkangia committed:
fix text

mkangia committed:
update dumper column headers

mkangia committed:
Use other constants

mkangia committed:
Merge branch 'master' into mk/location-reassignment-1

mkangia committed:
avoid indefinite spinner

mkangia committed:
add help link to avoid errors

mkangia committed:
stickler

mkangia committed:
dont fail if location missing

mkangia committed:
load only active descendants

mkangia committed:
Merge pull request #27056 from dimagi/mk/location-reassignment-4

Mk/location reassignment 4

mkangia committed:
better constant name

mkangia committed:
update col names as requested

mkangia committed:
inactive_objects removed

https://github.com/dimagi/commcare-hq/pull/27062/files

mkangia commented:
@esoergel merging this now to avoid merging a big elephant PR later..and deal with conflicts.
Feel free to drop feedback and i can take it in a follow up PR.
